ff/serving/provider/postgres.go

```diff
@@ -0,0 +1,501 @@
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (store *postgresOfflineStore) AsOfflineStore() (OfflineStore, error) {
+	return store, nil
+}
+
+// CreateResourceTable creates a new Resource table.
+// Returns a table if it does not already exist and stores the table ID in the resource index table.
+// Returns an error if the table already exists or if table is the wrong type.
+func (store *postgresOfflineStore) CreateResourceTable(id ResourceID) (OfflineTable, error) {
+	if err := id.check(Feature, Label); err != nil {
+		return nil, err
+	}
+
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if exists {
+		return nil, &TableAlreadyExists{id.Name, id.Variant}
+	}
+	//check if ctx is needed
+	tableName := store.getResourceTableName(id)
+	table, err := newPostgresOfflineTable(store.conn, tableName)
+	if err != nil {
+		return nil, err
+	}
+	return table, nil
+}
+
+func (store *postgresOfflineStore) GetResourceTable(id ResourceID) (OfflineTable, error) {
+	return store.getPostgresResourceTable(id)
+}
+
+func (store *postgresOfflineStore) getPostgresResourceTable(id ResourceID) (*postgresOfflineTable, error) {
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, &TableNotFound{id.Name, id.Variant}
+	}
+	return &postgresOfflineTable{
+		conn: store.conn,
+		ctx:  store.ctx,
+		name: store.getResourceTableName(id),
+	}, nil
+}
+
+func (store *postgresOfflineStore) CreateMaterialization(id ResourceID) (Materialization, error) {
+	if id.Type != Feature {
+		return nil, errors.New("only features can be materialized")
+	}
+	resTable, err := store.getPostgresResourceTable(id) // Need to clarify names
+	if err != nil {
+		return nil, err
+	}
+
+	matID := MaterializationID(id.Name)
+	matTableName := store.getMaterializationTableName(matID)
+	sanitizedTableName := sanitize(matTableName)
+	resTableName := sanitize(resTable.name)
+	tableCreateQry := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s AS (SELECT entity, value, ts FROM %s WHERE 1=2)", sanitizedTableName, resTableName)
+
+	_, err = store.conn.Exec(
+		context.Background(), tableCreateQry) // Set correct types
+	if err != nil {
+		return nil, err
+	}
+
+	materializeQry := fmt.Sprintf(
+		"INSERT INTO %s SELECT entity, value, ts FROM "+
+			"(SELECT entity, ts, value, row_number() OVER (PARTITION BY entity ORDER BY ts desc) "+
+			"AS rn FROM %s) t WHERE rn=1", sanitizedTableName, resTableName)
+
+	_, err = store.conn.Exec(context.Background(), materializeQry)
+	if err != nil {
+		return nil, err
+	}
+
+	return &postgresMaterialization{
+		id:        matID,
+		conn:      store.conn,
+		tableName: matTableName,
+	}, nil
+
+}
+func (store *postgresOfflineStore) GetMaterialization(id MaterializationID) (Materialization, error) {
+	getMatQry := fmt.Sprintf("SELECT DISTINCT (table_name) FROM information_schema.tables WHERE table_name LIKE 'materialization_%s%%'", string(id))
```

Comment:
given that you used sanitize to create the materialization table name, should the id pushed through getMaterializationTableName and sanitize be used for this query instead of hardcoding "materialization_" and using LIKE? What if a materialization matches another for the first n characters?

Reply:
Leftover from testing. Replaced with a prepared statement